src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -49,194 +47,96 @@ internal enum NullableFlowState : byte
-            }
-
-            if (a == NullableAnnotation.Annotated || b == NullableAnnotation.Annotated)
-            {
-                return NullableAnnotation.Annotated;
-            }
-
-            if (a == NullableAnnotation.Unknown || b == NullableAnnotation.Unknown)
-            {
-                return NullableAnnotation.Unknown;
-            }
-
-            if (a == NullableAnnotation.NotNullable || b == NullableAnnotation.NotNullable)
-            {
-                return NullableAnnotation.NotNullable;
-            }
-
-            return NullableAnnotation.NotAnnotated;
+            var x = _permute[(int)a];
+            var y = _permute[(int)b];
+            return (NullableAnnotation)_permute[(x > y) ? x : y];
         }
 
         /// <summary>
-        /// Join nullable flow states from distinct branches during flow analysis.
+        /// Meet two nullable annotations for computing the nullable annotation of a type parameter from upper bounds.
+        /// This uses the contravariant merging rules.
         /// </summary>
-        public static NullableFlowState JoinForFlowAnalysisBranches(this NullableFlowState selfState, NullableFlowState otherState)
+        public static NullableAnnotation Meet(this NullableAnnotation a, NullableAnnotation b)
         {
-            return (selfState == NullableFlowState.MaybeNull || otherState == NullableFlowState.MaybeNull)
-                ? NullableFlowState.MaybeNull : NullableFlowState.NotNull;
+            var x = _permute[(int)a];
+            var y = _permute[(int)b];
+            return (NullableAnnotation)_permute[(x < y) ? x : y];
         }
 
         /// <summary>
-        /// Meet two nullable annotations for computing the nullable annotation of a type parameter from upper bounds.
-        /// This uses the contravariant merging rules.
+        /// This permutation places the NullableFlowState values into the proper order to form a linear
+        /// lattice.  This permutation is a self inverse.
         /// </summary>
-        public static NullableAnnotation MeetForFixingUpperBounds(this NullableAnnotation a, NullableAnnotation b)
-        {
-            if (a == NullableAnnotation.NotNullable || b == NullableAnnotation.NotNullable)
-            {
-                return NullableAnnotation.NotNullable;
-            }
-
-            if (a == NullableAnnotation.NotAnnotated || b == NullableAnnotation.NotAnnotated)
-            {
-                return NullableAnnotation.NotAnnotated;
-            }
-
-            if (a == NullableAnnotation.Unknown || b == NullableAnnotation.Unknown)
-            {
-                return NullableAnnotation.Unknown;
-            }
-
-            if (a == NullableAnnotation.Nullable || b == NullableAnnotation.Nullable)
-            {
-                return NullableAnnotation.Nullable;
-            }
-
-            return NullableAnnotation.Annotated;
-        }
+        private static int[] _permute = new[] { 1, 0, 2 };
```

Comment:
>private static int[] _permute = new[] { 1, 0, 2 }; [](start = 8, length = 50)

This trick feels too clever (ie. hard to review). I think we should either write the Meet/Join in an expanded form, or we should re-order the enum entries.

Reply:
I would love to reorder, but we have plenty of hand-written IL with hardcoded attribute values with the enum values hardcoded.  That would be a fair amount of work to fix up the tests but I think worth it.  Opened [https://github.com/dotnet/roslyn/issues/33952](https://github.com/dotnet/roslyn/issues/33952) to track this.

---
In reply to: [263508874](https://github.com/dotnet/roslyn/pull/33927#discussion_r263508874) [](ancestors = 263508874)

Reply:
For what it's worth, my suggestion is to re-order the enum, but keep the IL that we emit as-is (otherwise we'll be causing unnecessary surprises to early adopters)

---
In reply to: [263598753](https://github.com/dotnet/roslyn/pull/33927#discussion_r263598753) [](ancestors = 263598753,263508874)

Reply:
@cston points out that reordering would be a breaking change.  It is really sad to be constrained by breaks at this juncture.

---
In reply to: [263598753](https://github.com/dotnet/roslyn/pull/33927#discussion_r263598753) [](ancestors = 263598753,263508874)